govcd/api.go

```diff
@@ -178,3 +179,122 @@ func checkResp(resp *http.Response, err error) (*http.Response, error) {
+}
+
+// Helper function creates request, run it, check response and do not expect any values from it.
+// pathUlr - request URL
+// requestType - HTTP method type
+// contentType - value to set for "Content-Type"
+// errorMessage - error message to return when error happen
+// payload - XML struct which will be marshalled and added as body/payload
+// E.g. client.ExecuteRequestWithoutResponse(catalogItemHREF.String(), http.MethodDelete, "", "error deleting Catalog item: %s", nil)
+func (client *Client) ExecuteRequestWithoutResponse(pathURL, requestType, contentType, errorMessage string, payload interface{}) error {
+
+	resp, err := executeRequest(pathURL, requestType, contentType, payload, client)
+	if err != nil {
+		return fmt.Errorf(errorMessage, err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf(errorMessage, err)
+	}
+
+	// The request was successful
+	return nil
+}
+
+// Helper function creates request, run it, check response and parse our interface from response.
+// pathUlr - request URL
+// requestType - HTTP method type
+// contentType - value to set for "Content-Type"
+// errorMessage - error message to return when error happen
+// payload - XML struct which will be marshalled and added as body/payload
+// out - structure to be used for unmarshalling xml
+// E.g. 	unmarshalledAdminOrg := &types.AdminOrg{}
+// client.ExecuteRequest(adminOrg.AdminOrg.HREF, http.MethodGet, "", "error refreshing organization: %s", nil, unmarshalledAdminOrg)
+func (client *Client) ExecuteRequest(pathURL, requestType, contentType, errorMessage string, payload, out interface{}) error {
+
+	if !isMessageWithPlaceHolder(errorMessage) {
+		return fmt.Errorf("error message has to include place holder for error")
+	}
+
+	resp, err := executeRequest(pathURL, requestType, contentType, payload, client)
+	if err != nil {
+		return fmt.Errorf(errorMessage, err)
+	}
+
+	if err = decodeBody(resp, out); err != nil {
+		return fmt.Errorf("error decoding response: %s", err)
+	}
+
+	err = resp.Body.Close()
+	if err != nil {
+		return fmt.Errorf(errorMessage, err)
+	}
+
+	// The request was successful
+	return nil
+}
+
+func executeRequest(pathURL, requestType, contentType string, payload interface{}, client *Client) (*http.Response, error) {
+	url, _ := url.ParseRequestURI(pathURL)
+
+	var req *http.Request
+	switch requestType {
+	case http.MethodPost, http.MethodPut:
+
+		marshaledXml, err := xml.MarshalIndent(payload, "  ", "    ")
+		if err != nil {
+			return &http.Response{}, fmt.Errorf("error marshalling xml data %v", err)
+		}
+		body := bytes.NewBufferString(xml.Header + string(marshaledXml))
+
+		req = client.NewRequest(map[string]string{}, requestType, *url, body)
+
+	default:
+		req = client.NewRequest(map[string]string{}, requestType, *url, nil)
+	}
+
+	if contentType != "" {
+		req.Header.Add("Content-Type", contentType)
+	}
+
+	return checkResp(client.Http.Do(req))
+}
+
+func isMessageWithPlaceHolder(message string) bool {
```

Comment:
Very nice implementation. I was thinking of using a regex for this purpose, but I see that this method is more effective.

Reply:
Thank you, it took some time to figure out how to catch missing placeholder at all